### ISSUE TITLE
`required=False` is not respected in DRF 2.x

### ISSUE THREAD

**Issue by hongquan** [tomchristie/django-rest-framework#3507]
This line of code: https://github.com/tomchristie/django-rest-framework/blob/2.4.8/rest_framework/serializers.py#L952

```python
        for field_name, field in self.fields.items():
            field_name = field.source or field_name
            if (
                field_name in exclusions
                and not field.read_only
                and (field.required or hasattr(instance, field_name))
                and not isinstance(field, Serializer)
            ):
                exclusions.remove(field_name)
```

I have a _User_ model, with `password` being one of the fields. Now I want to update _User_ via API. I define `UserSerializer` whose `password` field is `required=False`.

Because of the code above, when I submit data with empty `password` field, my data cannot pass the validation:

```python
 {'password': [u'This field cannot be blank.']}
```
The expression

```python
(field.required or hasattr(instance, field_name))
```
is `True` because `field.required` is `False` and `hasattr(instance, field_name)` is `True`.

Testing with
djangorestframework==2.4.8
Django==1.4.22

**Comment by xordoquy** [comment]
We don't maintain 2.x any more although we may merge pull request that fixes issues.
Note that you are probably confusing required with blank parameter. If the field isn't provided (`required=False`), there's no reason `hasattr(instance, field_name)` is True.

**Comment by hongquan** [comment]
You are wrong. I put `print` to debug the value as following:

```python
for field_name, field in self.fields.items():
            field_name = field.source or field_name
            if (
                field_name in exclusions
                and not field.read_only
                and (field.required or hasattr(instance, field_name))
                and not isinstance(field, Serializer)
            ):
                exclusions.remove(field_name)
                if field_name == 'password':
                    print field.required, hasattr(instance, field_name)
```

And here is test's output

```
-------------------- >> begin captured stdout << ---------------------
False True

--------------------- >> end captured stdout << ----------------------
```

**Comment by xordoquy** [comment]
Sorry, I meant that there are no reason the `hasattr(instance, field_name)` could be False when the field is provided.

**Comment by hongquan** [comment]
Then, how to workaround this? I already have:

```python
password = serializers.CharField(required=False)
```

in the serializer. I cannot make change in _User_ model because it is from `django.contrib.auth`.

**Comment by xordoquy** [comment]
As I said earlier we don't support 2.4 any longer. I can't afford to spent free time on a dead code line and I think we fixed that in 3.0 by improving how required / blank / true.

**Comment by hongquan** [comment]
Thank you for your reply. I hope I can upgrade to 3.0 someday.
